Add Number Normalisation for SpeechT5

tanaymeh opened the pull request:
# What does this PR do?
This PR will add number normalisation for SpeechT5. Currently, SpeechT5 cannot read numbers, as described in #23480.

Fixes #23480

# Cases covered so far
- [x] Integer number normalisation
- [x] Floating point number normalisation
- [x] Currency normalisation
	- [x] US Dollar `$`
	- [x] cents `¢`
	- [x] Euro `€`
	- [x] Sterling Pound `£`
	- [x] Yen `¥`
	- [x] Other major currency symbols
- [x] Percents `%` 
- [x] Negative numbers (starting with `-`)
- [x] Numbers from 0 to Decillion
- [x] Any combination of those stated and ticked above

## Who can review?
@sanchit-gandhi

tanaymeh committed:
add: NumberNormalizer works for integers, floats, common currencies, negative numbers and percentages

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_25447). All of your documentation changes will be reflected on that endpoint.

sgugger commented:
cc @sanchit-gandhi

tanaymeh committed:
fix: renamed number normalizer class and added normalization to SpeechT5Processor

tanaymeh committed:
fix: restyled with black and ruff, should pass code quality tests

sanchit-gandhi commented on src/transformers/models/speecht5/number_normalizer.py:
```diff
@@ -0,0 +1,198 @@
+import re
+
+
+class EnglishNumberNormalizer:
```
This is looking super comprehensive already! Thanks for listing all the normalisations applied in the PR comment - are there any big ones that we're missing from Whisper?

tanaymeh replied:
Thanks! To my knowledge, all cases are covered from Whisper

sanchit-gandhi commented on src/transformers/models/speecht5/number_normalizer.py:
```diff
@@ -0,0 +1,198 @@
+            integer_part = integer_part.replace("%", "")
+            decimal_part = decimal_part.replace("%", "")
+
+        print(integer_part, decimal_part)
```
No need for any print statements! Might have been left in from debugging

sanchit-gandhi commented on src/transformers/models/speecht5/number_normalizer.py:
```diff
@@ -0,0 +1,198 @@
+
+        print(integer_part, decimal_part)
+
+        # integer_part = re.sub(r"\D", "", integer_part)
```
likewise this comment

sanchit-gandhi commented on src/transformers/models/speecht5/number_normalizer.py:
```diff
@@ -0,0 +1,198 @@
+        integer_part = integer_part.zfill(3 * ((len(integer_part) - 1) // 3 + 1))
+
+        parts = []
+        units = [
```
Can we not use `self.thousands` here?

tanaymeh replied:
Yes! I actually wrote this function separately while working on it so I didn't integrate it well with the class. Fixed now!

sanchit-gandhi commented on src/transformers/models/speecht5/number_normalizer.py:
```diff
@@ -0,0 +1,198 @@
+
+        # Define a dictionary to map currency symbols to their names
+        currency_symbols = {
+            "$": " dollars",
```
Should these have the extra space before?

tanaymeh replied:
Yes, it is to ensure that `$12` becomes `twelve dollars` and not `twelvedollars`. Adding a space in the symbol to name mapping is easier and less complicated than doing it in the format string which is returned.

sanchit-gandhi replied:
Cool cool - thanks for clarifying! Good with me to do it this way

sanchit-gandhi reviewed (commented):
Looks very comprehensive already @tanaymeh! Wonder if we could add a few tests to [`test_tokenization_speech_t5.py`](https://github.com/huggingface/transformers/blob/b42010bb1d3cbf262d27e0a328661885be46dfdb/tests/models/speecht5/test_tokenization_speecht5.py#L32) to confirm we get the expected behaviour? E.g. including tests so that we cover each of the normalisation cases and confirm they work as expected (you can test multiple parts in single tests if you want, e.g. numbers and currency in one go)

tanaymeh commented on src/transformers/models/speecht5/number_normalizer.py:
```diff
@@ -0,0 +1,198 @@
+
+        return " ".join(reversed(parts))
+
+    def convert(self, n):
```
Will do!

tanaymeh commented on src/transformers/models/speecht5/processing_speecht5.py:
```diff
@@ -15,6 +15,7 @@
+from .number_normalizer import EnglishNumberNormalizer
 
 
 class SpeechT5Processor(ProcessorMixin):
```
That makes sense. Just want to confirm, the right place for `if self.normalize` check should be in the `_tokenize()` function right?

https://github.com/huggingface/transformers/blob/b42010bb1d3cbf262d27e0a328661885be46dfdb/src/transformers/models/speecht5/tokenization_speecht5.py#L143

tanaymeh committed:
fix: moved normalization to tokenizer and other small changes to normalizer

tanaymeh committed:
add: test for normalization and changed the existing full tokenizer test

tanaymeh committed:
fix: tokenization tests now pass, made changes to existing tokenization where normalization is covered; added normalize arg to func signature

tanaymeh commented:
@sanchit-gandhi I added a single test which tests all the covered cases. I also fixed an existing test in the `tests/models/speecht5/test_tokenization_speecht5.py` file which previously was just passing the number '92000' around without normalization but now it would normalize it (see below)

https://github.com/tanaymeh/transformers/blob/afa90898b93c58e712d6d7e3726a66d964282930/tests/models/speecht5/test_tokenization_speecht5.py#L164-L183

sanchit-gandhi commented on tests/models/speecht5/test_tokenization_speecht5.py:
```diff
@@ -52,12 +52,24 @@ def get_input_output_texts(self, tokenizer):
         output_text = "this is a test"
         return input_text, output_text
 
+    def get_numeric_input_output_texts(self):
```
Nice - thanks for checking all cases!

ArthurZucker replied:
I am guessing that `"8,000"` is for American normalizer? Or should it also be included?

tanaymeh replied:
Sorry, I didn't quite understand what you were referring to @ArthurZucker

ArthurZucker replied:
Just saying that some might write `8000` as `8,000` and wondering if this should also be accounted for?

tanaymeh replied:
Got it! Adding the below line in the next commit which deals with this issue and normalizes any number with comma separation.

```python
text = re.sub(r"(\d+,\d+)", lambda match: match.group(1).replace(",", ""), text)
```

sanchit-gandhi commented on tests/models/speecht5/test_tokenization_speecht5.py:
```diff
@@ -52,12 +52,24 @@ def get_input_output_texts(self, tokenizer):
+    def test_tokenizer_normalization(self):
+        tokenizer = self.get_tokenizer()
+        input_text, output_text = self.get_numeric_input_output_texts()
+        input_ids = tokenizer.convert_tokens_to_ids(tokenizer.tokenize(input_text))
```
Could we make this a bit more clear by setting `normalize=True` for the one we want to normalize?

Is the test also maybe a bit better if we check that the output of the tokenizer is equal to the expected text? E.g.
```python
input_text, expected_text = self.get_numeric_input_output_texts()
input_ids = tokenizer.encode(input_text, normalize=True)
output_text = tokenizer.decode(input_ids, skip_special_tokens=True)
self.assertListEqual(output_text, tokenizer.encode)
```

sanchit-gandhi commented on tests/models/speecht5/test_tokenization_speecht5.py:
```diff
@@ -153,20 +165,20 @@ def test_full_tokenizer(self):
             tokens,
             # fmt: off
-            [SPIECE_UNDERLINE, 'I', SPIECE_UNDERLINE, 'w', 'a', 's', SPIECE_UNDERLINE, 'b', 'o', 'r', 'n', SPIECE_UNDERLINE, 'i', 'n', SPIECE_UNDERLINE, '92000', ',', SPIECE_UNDERLINE, 'a', 'n', 'd', SPIECE_UNDERLINE, 't', 'h', 'i', 's', SPIECE_UNDERLINE, 'i', 's', SPIECE_UNDERLINE, 'f', 'a', 'l', 's', 'é', '.']
+            [SPIECE_UNDERLINE, 'I', SPIECE_UNDERLINE, 'w', 'a', 's', SPIECE_UNDERLINE, 'b', 'o', 'r', 'n', SPIECE_UNDERLINE, 'i', 'n', SPIECE_UNDERLINE, 'n', 'i', 'n', 'e', 't', 'y', SPIECE_UNDERLINE, 't', 'w', 'o', SPIECE_UNDERLINE, 't', 'h', 'o', 'u', 's', 'a', 'n', 'd', ',', SPIECE_UNDERLINE, 'a', 'n', 'd', SPIECE_UNDERLINE, 't', 'h', 'i', 's', SPIECE_UNDERLINE, 'i', 's', SPIECE_UNDERLINE, 'f' ...
```
Thanks for updating these tests

sanchit-gandhi reviewed (approved):
Thanks for adding the tests! LGTM - would just request that we change the default behaviour to not normalise, and maybe add a line in the SpeechT5 tokenizer docs to highlight to users that they can use this great feature?

tanaymeh commented on src/transformers/models/speecht5/tokenization_speecht5.py:
```diff
@@ -97,6 +100,7 @@ def __init__(
         eos_token="</s>",
         unk_token="<unk>",
         pad_token="<pad>",
+        normalize=True,
```
That makes sense, having it on `normalize=True` may break existing code where this is used. I have now set it to False.

tanaymeh commented:
I have a rather dumb doubt, @sanchit-gandhi: I had added the information about the `normalize` argument in the docstring of the `SpeechT5Tokenizer` class ([here](https://github.com/tanaymeh/transformers/blob/afa90898b93c58e712d6d7e3726a66d964282930/src/transformers/models/speecht5/tokenization_speecht5.py#L68-L69)). Should I add it somewhere else to be shown in the docs? Doesn't the doc builder builds the docstrings and renders them in the docs automatically?

Thanks!

tanaymeh committed:
fix: changed default normalize setting to False, modified the tests a bit

ArthurZucker commented on src/transformers/models/speecht5/number_normalizer.py:
```diff
@@ -0,0 +1,186 @@
+# coding=utf-8
+# Copyright 2023 The Facebook Inc. and The HuggingFace Inc. team. All rights reserved.
```
Is speecht5 not from microsoft?

tanaymeh replied:
Yes, Apologies, I copy-pasted the comments at the top from another file and forgot to change it.

ArthurZucker commented on src/transformers/models/speecht5/number_normalizer.py:
```diff
@@ -0,0 +1,186 @@
+            "decillion",
+        ]
+
+        # Define a dictionary to map currency symbols to their names
+        self.currency_symbols = {
```
The list only include certain currencies (of course!) which short list does it correspond to?

ArthurZucker replied:
Not at all! Can you just mention in a comment `# top x most popular currencies according to xxxx ` can just help us know why we went for this list!

ArthurZucker commented on src/transformers/models/speecht5/tokenization_speecht5.py:
```diff
@@ -142,6 +148,8 @@ def __setstate__(self, d):
 
     def _tokenize(self, text: str) -> List[str]:
         """Take as input a string and return a list of strings (tokens) for words/sub-words"""
+        if self.normalize:
+            text = self._normalize(text)
```
This should go in the `prepare_inputs_for_tokenization`. This way it could support normalizing on the fly with kwargs, and only normalize once if there are special tokens (instead of normalizing each substrings)

tanaymeh replied:
Makes sense, will do that!

tanaymeh replied:
@ArthurZucker Should I make a new function called `prepare_inputs_for_tokenization`?

ArthurZucker replied:
It's actually `prepare_for_tokenization` which should be overloaded!

tanaymeh replied:
Gotcha. This is what I am doing currently @ArthurZucker;

```python
# ... rest of the code in Tokenizer class
    @property
    def normalizer(self):
        return self.normalizer

    @normalizer.setter
    def normalizer(self):
        self.normalizer = EnglishNumberNormalizer()
# ... other code

    def prepare_for_tokenization(self, text, is_split_into_words=False, **kwargs):
        normalize = kwargs.pop("normalize", self.normalize)
        if is_split_into_words:
            text = " " + text
        if normalize:
            text = self.normalizer(text)
        return (text, kwargs)
```

Am I on the right track / any suggestions?

ArthurZucker commented on src/transformers/models/speecht5/tokenization_speecht5.py:
```diff
@@ -203,3 +211,11 @@ def save_vocabulary(self, save_directory: str, filename_prefix: Optional[str] =
 
         return (out_vocab_file,)
+
+    def _normalize(self, text: str):
+        """
+        Normalizes a given string using the 'EnglishNumberNormalizer' class, which converts numeric elements in english
+        text.
+        """
+        normalizer = EnglishNumberNormalizer()
+        return normalizer(text)
```
No sure we need a method for this, re-initializing the normalizer on each tokenization.

tanaymeh replied:
I took the inspiration from Whisper. Instead, should I now initialize the normalizer once in the `__init__()` function and use it directly?

ArthurZucker reviewed (commented):
Thanks a lot for working on this! Left a few small nits but otherwise looks good!

tanaymeh committed:
fix: added support for comma separated numbers, tokenization on the fly with kwargs and normalizer getter setter funcs

tanaymeh commented:
@ArthurZucker I added your suggested changes ([here](https://github.com/tanaymeh/transformers/blob/682bee8d517398b8b9840808ff3a3bb106f6b9fd/src/transformers/models/speecht5/tokenization_speecht5.py#L126-L132) and [here](https://github.com/tanaymeh/transformers/blob/682bee8d517398b8b9840808ff3a3bb106f6b9fd/src/transformers/models/speecht5/tokenization_speecht5.py#L138-L146)).

I also added support for comma-separated numbers [here](https://github.com/tanaymeh/transformers/blob/682bee8d517398b8b9840808ff3a3bb106f6b9fd/src/transformers/models/speecht5/number_normalizer.py#L185-L186).

Please let me know if I missing something or need to change something.

ArthurZucker reviewed (approved):
Looks good to me! Thanks for iterating 🤗

tanaymeh commented:
Thanks for your review @sanchit-gandhi and @ArthurZucker. Waiting for review from @ylacombe!

ramkrishna757575 commented:
I am rather new to python and transformers.
Can someone please provide me documentation for this? I am doing the following, but I am not sure where I can pass the normalize parameter.
Currently the audio being generated skips all the numbers given in the input text.

```
import sounddevice as sd
import torch
from datasets import load_dataset
from transformers import SpeechT5Processor, SpeechT5HifiGan, SpeechT5ForTextToSpeech

processor = SpeechT5Processor.from_pretrained("microsoft/speecht5_tts")
embeddings_dataset = load_dataset("Matthijs/cmu-arctic-xvectors", split="validation")
speaker_embeddings = torch.tensor(embeddings_dataset[7306]["xvector"]).unsqueeze(0)
device = "cuda:0" if torch.cuda.is_available() else "cpu"


def synthesise(text):
    inputs = processor(text=text, return_tensors="pt")
    vocoder = SpeechT5HifiGan.from_pretrained("microsoft/speecht5_hifigan").to(device)
    model = SpeechT5ForTextToSpeech.from_pretrained("microsoft/speecht5_tts").to(device)
    speech = model.generate_speech(
        inputs["input_ids"].to(device), speaker_embeddings.to(device), vocoder=vocoder
    )
    return speech.cpu()


def playback_speech(text):
    audio = synthesise(text)
    sd.play(audio.numpy(), samplerate=16000)
    sd.wait()


playback_speech("Numbers from one through ten : 1 2 3 4 5 6 7 8 9 10")
```

tanaymeh commented:
@ramkrishna757575 You need to turn `normalize=True` in the `processor`. By default, the number normalization is turned off to ensure backwards compatibility since it's a breaking change.

Inside the `synthesise()` function, you can do:
```python
# ... Rest of the code
def synthesise(text):
    inputs = processor(text=text, return_tensors="pt", normalize=True)
# ... Rest of the code
```

ramkrishna757575 commented:
> 

Thanks for the quick reply @tanaymeh.
I tried it, but I got this error:

`Keyword arguments {'normalize': True} not recognized.`

I checked the version of `transformers` that I have installed and it is currently `4.32.0`

tanaymeh commented:
@ramkrishna757575 About the numbers not being spelt, I am not sure what the reason is. It is certainly not with the processor or tokenizer since both are working fine. See the attached screenshot below.

<img width="1071" alt="Screenshot 2023-08-26 at 7 17 14 PM" src="https://github.com/huggingface/transformers/assets/26519539/64463bc8-f8e9-42e6-adc8-cfca7286d569">

tanaymeh commented:
@ramkrishna757575 You need to install the latest version of the transformers (that is not yet released), using:
`pip install git+https://github.com/huggingface/transformers`, in this, you will be able to use this setting.

ramkrishna757575 commented:
ok..thanks @tanaymeh .
I'll try this out.
Will post my findings later once I'm done.

sanchit-gandhi commented:
Thanks for helping out here @tanaymeh! Here's the parameter in the docs (on `main`) @ramkrishna757575: https://huggingface.co/docs/transformers/main/en/model_doc/speecht5#transformers.SpeechT5Tokenizer.normalize

ramkrishna757575 commented:
> Thanks for helping out here @tanaymeh! Here's the parameter in the docs (on `main`) @ramkrishna757575: https://huggingface.co/docs/transformers/main/en/model_doc/speecht5#transformers.SpeechT5Tokenizer.normalize

got it...thanks @sanchit-gandhi
